src/components/MessageList/hooks/VirtualizedMessageList/useVirtualizedMessageListComponents.tsx

```diff
@@ -0,0 +1,200 @@
+
+      return (
+        <Message
+          additionalMessageInputProps={additionalMessageInputProps}
+          autoscrollToBottom={virtuosoRef.current?.autoscrollToBottom}
+          closeReactionSelectorOnClick={closeReactionSelectorOnClick}
+          customMessageActions={customMessageActions}
+          endOfGroup={endOfGroup}
+          firstOfGroup={firstOfGroup}
+          groupedByUser={groupedByUser}
+          lastReceivedId={lastReceivedMessageId}
+          message={message}
+          Message={MessageUIComponent}
+          messageActions={messageActions}
+          readBy={ownMessagesReadByOthers[message.id] || []}
+        />
+      );
+    },
+    [
+      additionalMessageInputProps,
+      closeReactionSelectorOnClick,
+      customMessageActions,
+      customMessageRenderer,
+      messageActions,
+      shouldGroupByUser,
+      prependOffset,
+    ],
+  );
+
+  const virtuosoComponents: Partial<
+    Components<VirtuosoContext<StreamChatGenerics>>
+  > = useMemo(() => {
+    // using 'display: inline-block'
+    // traps CSS margins of the item elements, preventing incorrect item measurements
+    const Item: Components<VirtuosoContext<StreamChatGenerics>>['Item'] = ({
+      context,
+      ...props
+    }) => {
+      if (!context) return <></>;
+
+      const streamMessageIndex =
+        props['data-item-index'] + context.numItemsPrepended - prependOffset;
+      const message = context.processedMessages[streamMessageIndex];
+      const groupStyles: GroupStyle = context.messageGroupStyles[message.id] || '';
+
+      return (
+        <div
+          {...props}
+          className={
+            context?.customClasses?.virtualMessage ||
+            clsx('str-chat__virtual-list-message-wrapper str-chat__li', {
+              [`str-chat__li--${groupStyles}`]: groupStyles,
+            })
+          }
+        />
+      );
+    };
+
+    const EmptyPlaceholder: Components<
+      VirtuosoContext<StreamChatGenerics>
+    >['EmptyPlaceholder'] = () => (
+      <>
+        {EmptyStateIndicator && (
+          <EmptyStateIndicator listType={threadList ? 'thread' : 'message'} />
+        )}
+      </>
+    );
+
+    const Header: Components<VirtuosoContext<StreamChatGenerics>>['Header'] = () =>
+      loadingMore && !!LoadingIndicator ? (
+        <div className='str-chat__virtual-list__loading'>
+          <LoadingIndicator size={20} />
+        </div>
+      ) : (
+        head || null
+      );
+
+    const Footer: Components<VirtuosoContext<StreamChatGenerics>>['Footer'] = () =>
+      TypingIndicator ? <TypingIndicator avatarSize={24} /> : <></>;
+
+    return {
+      EmptyPlaceholder,
+      Footer,
+      Header,
+      Item,
+      ...virtuosoComponentsFromProps,
+    };
+  }, [
+    EmptyStateIndicator,
+    LoadingIndicator,
+    TypingIndicator,
+    loadingMore,
+    head,
+    threadList,
+    virtuosoComponentsFromProps,
```

Comment:
I have a feeling that this may re-introduce this bug: https://github.com/GetStream/stream-chat-react/pull/2067
as it can break the memoization, in case someone uses this:
```
<VirualizedMessageList additionalVirtuosoProps={{ components: { Footer: MyFooter }}} />
```

Can you please check that?

Reply:
Checked, whether the list rerenders:

1. add video
2. play video
3. add new messages

The video kept playing.

I have also added a current timestamp to the `components` objects to observe, whether the `components` object would change (if the timestamp changes). The timestamp does not change even when scrolling the list.

```
< VirualizedMessageList additionalVirtuosoProps={{ components: { Footer: MyFooter, d: new Date().toISOString() }}} />
```